lib/auth/password.go

```diff
@@ -358,18 +385,22 @@ func (s *Server) changePasswordWithToken(ctx context.Context, req ChangePassword
 	return user, nil
 }
 
-func (s *Server) changeUserSecondFactor(req ChangePasswordWithTokenRequest, ResetPasswordToken services.ResetPasswordToken) error {
-	username := ResetPasswordToken.GetUser()
+func (s *Server) changeUserSecondFactor(req ChangePasswordWithTokenRequest, token services.ResetPasswordToken) error {
+	username := token.GetUser()
 	cap, err := s.GetAuthPreference()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
 	ctx := context.TODO()
-	switch cap.GetSecondFactor() {
-	case teleport.OFF:
+	secondFactor := cap.GetSecondFactor()
+	if secondFactor == constants.SecondFactorOff {
 		return nil
-	case teleport.OTP, teleport.TOTP, teleport.HOTP:
+	}
+	if req.SecondFactorToken != "" {
+		if secondFactor == constants.SecondFactorU2F {
+			return trace.BadParameter("user %q sent a TOTP token during password reset but cluster only allows U2F for second factor", username)
```

Comment:
Should this be more generic like it is in the error below (as we also technically support HOTP, I think?)

```suggestion
			return trace.BadParameter("user %q sent an OTP token during password reset but cluster only allows U2F for second factor", username)
```